---
title: "Secreted protein prediction"
author: "Xabier Vázquez-Campos"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: 4
---

## Workflow scheme
```{r, echo=FALSE}
library(DiagrammeR)
grViz("
      digraph nicegraph {
      
      # graph, node, and edge definitions
      graph [compound = true]
      
      # program nodes
      node [fontname = Helvetica, shape = rectangle, fontsize = 12, fixedsize = true,
      color = darkslategray, width = 1.3]
      A[label='SignalP']
      B[label='Phobius']
      C[label='TMHMM']
      D[label='GPS-Lipid']
      E[label='KohGPI']
      F[label='TargetP']
      G[label='selectSeqs.pl']
      H[label='EffectorP']
      I[label='ApoplastP']
      
      # protein collection nodes
      node [fontname = Helvetica, shape = oval, fontsize = 10, color = darkslategray, width = 1.6]
      a[label='all proteins']
      b[label='unannotated\nproteins', style = filled, fillcolor = salmon1]
      c[label='signal peptide\nconsensus']
      d[label='Phobius\ndiscards', style = filled, fillcolor = salmon1]
      e[label='SignalP\ndiscards', style = filled, fillcolor = salmon1]
      f[label='SignalP≠Phobius', style = filled, fillcolor = salmon1]
      g[label='prots w/rogue\nTM domains', style = filled, fillcolor = salmon1]
      h[label='prots not targeting\nSec path', style = filled, fillcolor = salmon1]
      i[label='prots with\nGPI-anchor patterns', style = filled, fillcolor = salmon1]
      j[label='Putative\nSecreted Proteins', style = filled, fillcolor = green1]
      k[label='prots with\nprenyl/lipid mods', style = filled, fillcolor = salmon1]
      l[label='CSSP', style = filled, fillcolor = green3, fontsize = 18, penwidth = 4]
      m[label='Heavy\nSecreted Proteins', style = filled, fillcolor = green1]
      n[label='CSEP', style = filled, fillcolor = green1]
      o[label='Non-effector\nCSSP', style = filled, fillcolor = salmon1]
      p[label='Apoplast\nEffectors', style = filled, fillcolor = green1]
      q[label='Non-apoplast\nEffectors', style = filled, fillcolor = green1]
      
      # processing edges
      edge [color = grey, arrowhead = normal, arrowsize = 0.5, arrowtail = none, fontsize = 8]
      a -> A[label = 'with\nSignalP\ndomain']
      a -> B[label = 'with\nPhobius\ndomain']
      A -> c[label='noTM']
      B -> c[label='signal\npeptide']
      c -> C
      C -> F[label='No']
      F -> E[label='S']
      E -> D[label='No/\nundet']
      D -> j[label='No']
      j -> G[label='size-filtering']
      G -> l[label='≤300 aa']
      G -> m[label='>300 aa']
      l -> H
      H -> n[label='yes']
      n -> I
      I -> p[label='Apoplast']
      I -> q[label='Not apoplast']
      
      # discarding edges
      edge [color = grey, arrowhead = diamond, arrowtail = none, fontsize = 6]
      a -> b[label='no domain']
      B -> d[label='cyt dom\nand/or\ntransmemb dom', y = middle]
      A -> e[label='TM']
      c -> f
      C -> g[label='Yes']
      F -> h[label='other']
      E -> i[label='Yes']
      D -> k[label='Yes']
      H -> o[label='No']
      
      
      subgraph {
      rank = same; a; b;
      }
      subgraph {
      rank = same; d; B;
      }
      subgraph {
      rank = same; c; f;
      }
      subgraph {
      rank = same; e; A;
      }
      subgraph {
      rank = same; C; g;
      }
      subgraph {
      rank = same; F; h;
      }
      subgraph {
      rank = same; E; i;
      }
      subgraph {
      rank = same; D; k;
      }
      subgraph {
      rank = same; H; o;
      }
      }")
```

## Requirements
### Software
* [`selectSeqs.pl`](http://raven.iab.alaska.edu/~ntakebay/teaching/programming/perl-scripts/selectSeqs.pl)
* [KohGPI 1.5](http://gpi.unibe.ch/kohgpi-1.5.tar.gz)
* [GPS-Lipid 1.0](http://lipid.biocuckoo.org/download.php)

### Files
From Maker's annotation protocol:

* `${MYGENOME}.all.maker.proteins.renamed.iprsout.tsv`
* `${MYGENOME}.all.maker.proteins.renamed.func.fasta`


## Protocol

First, create a sorted IPRS output file:
```{bash sort, eval=FALSE}
cd annotation/
mkdir -p cssp
cd cssp
sort ../${MYGENOME}.all.maker.proteins.renamed.iprsout.tsv > ${MYGENOME}.all.maker.proteins.renamed.iprsout.tsv.sort
```

### Phobius
Get proteins with signal peptide from Phobius
```{bash phobius, eval=FALSE}
grep "Phobius"$'\t'"SIGNAL_PEPTIDE" ${MYGENOME}.all.maker.proteins.renamed.iprsout.tsv.sort |\
cut -f1 | sort | uniq > ${MYGENOME}_phobius_signalpep.txt

grep "Phobius"$'\t'"CYTOPLASMIC_DOMAIN\|Phobius"$'\t'"TRANSMEMBRANE" ${MYGENOME}.all.maker.proteins.renamed.iprsout.tsv.sort |\
cut -f1 | sort | uniq > ${MYGENOME}_phobius_TMorCYT.txt

comm -23 ${MYGENOME}_phobius_signalpep.txt ${MYGENOME}_phobius_TMorCYT.txt > ${MYGENOME}_phobius_signalpep_extracel.txt
```
> Proteins with signal peptides as predicted by Phobius can also have cytoplasmic and/or transmembrane domains, so we need to remove these ones from the list of proteins with signal peptides.  

### SignalP
Do the same for SignalP:
```{bash signalp, eval=FALSE}
grep "SignalP-noTM"$'\t' ${MYGENOME}.all.maker.proteins.renamed.iprsout.tsv.sort |\
cut -f1 | sort | uniq > ${MYGENOME}_signalp_noTM.txt
```
> SignalP only gives a single annotation per protein, so a simple filtering is enough.  

Get proteins suggested by both Phobius and SignalP:
```{bash consensus, eval=FALSE}
comm -12 ${MYGENOME}_phobius_signalpep_extracel.txt ${MYGENOME}_signalp_noTM.txt > ${MYGENOME}_merged_signals.txt
```

### TMHMM
Remove those with rogue transmembrane domains usign TMHMM:
```{bash tmhmm, eval=FALSE}
grep "TMHMM"$'\t' ${MYGENOME}.all.maker.proteins.renamed.iprsout.tsv.sort |\
cut -f1 | sort | uniq > ${MYGENOME}_tmhmm.txt

comm -23 ${MYGENOME}_merged_signals.txt ${MYGENOME}_tmhmm.txt > ${MYGENOME}_merged_signals.notm.txt
```

Get the seqs for those prots:
```{bash, eval=FALSE}
perl ~/bin/selectSeqs.pl -pf ${MYGENOME}_merged_signals.notm.txt ../${MYGENOME}.all.maker.proteins.renamed.func.fasta > ${MYGENOME}_merged_signals.notm.fasta
```
  
### TargetP
Submit `${MYGENOME}_merged_signals.notm.fasta` to [TargetP webserver](http://www.cbs.dtu.dk/services/TargetP/). You may need to chop the file into chunks due to the server limitations (2,000 sequences and 200,000 amino acids per submission; each sequence not more than 4,000 amino acids).
  
  
Copy the output to plain text files and recover the list of proteins assigned to the secretory pathway (and get their sequences):
```{bash, eval=FALSE}
grep "   S    " ${MYGENOME}_merged.targetp.txt |\
cut -f1 -d " " | sort | uniq > ${MYGENOME}_merged.targetp.list.txt

perl ~/selectSeqs.pl -pf ${MYGENOME}_merged.targetp.list.txt ${MYGENOME}_merged_signals.notm.fasta > ${MYGENOME}_merged_signals.notm.targetp.fasta
```

### Post-translational modifications
Now we search for signals indicating post-translational modifications (PTM) that would anchor the proteins to the cellular membrane without having a protein membrane domain. These modifications include GPI-lation (glycosylphosphatidylinositolation), prenylation (farnesylation and geranylgeranylation) and lipidation (palmitoylation and myristoilation).

#### KohGPI
For the detection of GPI-lation sites, we use KohGPI:
First we need to convert the fasta file to one-liner fasta:
```{bash, eval=FALSE}
cd ~/kohgpi-1.5
./pregpi ${MYGENOME}_merged_signals.notm.targetp.fasta
```
This will create a copy of `${MYGENOME}_merged_signals.notm.targetp.fasta` as `${MYGENOME}_merged_signals.notm.targetp.fasta.old`, while the output keeps the name.  
Now the real thing:
```{bash, eval=FALSE}
cd ~/kohgpi-1.5
./kohgpi ${MYGENOME}_merged_signals.notm.targetp.fasta
```
> If you get this message: `map not found`, you need to retrain KohGPI. See `README` in the KohGPI folder. You may need to execute KohGPI from its folder to avoid this.

> The output is a bunch of files named as `${MYGENOME}_merged_signals.notm.targetp.fasta.xxx`, for xxx equal log, neg (negative), pos (positive), png (mapping image) and und (undetermined). We filter out pos (positives), although if you want to be more conservative, keep just the strictly negatives and get rid of any uncertain match:

```{bash, eval=FALSE}
grep ">" ${MYGENOME}_merged_signals.notm.targetp.fasta.pos |\
cut -f1 -d " " | sed 's/>//g' > ${MYGENOME}_merged_signals.notm.targetp.gpi.list.txt

perl ~/selectSeqs.pl -pvf ${MYGENOME}_merged_signals.notm.targetp.gpi.list.txt ${MYGENOME}_merged_signals.notm.targetp.fasta > ${MYGENOME}_merged_signals.notm.targetp.nogpi.fasta
```

#### GPS-Lipid
Now for the rest of the PTMs with GPS-Lipid, unfortunately, there is no command-line option, so:

* Start GPS-Lipid `~/GPS-Lipid/GPS-Lipid\ 1.0`.
* Select all PTM models and set the threshold to high to minimize false positives.
* Load sequence file, `${MYGENOME}_merged_signals.notm.targetp.nogpi.fasta`.
* Click submit.
* Export results to ${MYGENOME}_merged_signals.notm.targetp.nogpi.gps.tbl. This table contains only positive matches.

```{bash, eval=FALSE}
tail -n+2 ${MYGENOME}_merged_signals.notm.targetp.nogpi.gps.tbl | cut -f1 -d " " | sort | uniq > ${MYGENOME}_merged_signals.notm.targetp.nogpi.gps.list.txt

perl ~/selectSeqs.pl -pvf ${MYGENOME}_merged_signals.notm.targetp.nogpi.gps.list.txt ${MYGENOME}_merged_signals.notm.targetp.nogpi.fasta > ${MYGENOME}_merged_signals.notm.targetp.nogpi.nogps.fasta
```
> Final count, total secreted proteins = 495

### Finish it!
Lastly, get the proteins ≤300 amino acids (and we keep also the big ones, just for housekeeping):
```{bash, eval=FALSE}
grep ">" ${MYGENOME}_merged_signals.notm.targetp.nogpi.nogps.fasta |\
cut -f1 -d " " | sed 's/>//g' > ${MYGENOME}_putative_secreted_proteins.list.txt

perl ~/selectSeqs.pl -pf ${MYGENOME}_putative_secreted_proteins.list.txt ../${MYGENOME}.all.maker.proteins.renamed.func.fasta > ${MYGENOME}_putative_secreted_proteins.fasta

perl ~/bin/selectSeqs.pl -l -300 ${MYGENOME}_putative_secreted_proteins.fasta > ${MYGENOME}_putative_small_secreted_proteins.fasta

perl ~/bin/selectSeqs.pl -l 301- ${MYGENOME}_putative_secreted_proteins.fasta > ${MYGENOME}_putative_heavy_secreted_proteins.fasta
```
> Final count, CSSP = 393  
> Final count, CHSP = 102